kpatch-build: fix building of livepatch-based patches

euspectre opened the pull request:
kpatch-build currently requires Module.symvers for the Kpatch core
module unconditionally and fails if it is not found. This does not allow
using kpatch-build to prepare livepatch-based patches.

This patch fixes the problem.

Signed-off-by: Evgenii Shatokhin <eshatokhin@virtuozzo.com>

flaming-toast commented:
Thanks for the patch @euspectre. This has indeed been a slightly annoying issue for a while. I think it might be more helpful though, to have this implemented as a command line option/switch for kpatch-build. Something like `-coremod` or `-nocoremod`, that would indicate whether we are building with the core kpatch module or building a livepatch module. This way we won't be dependent on the host system which could be building patch modules intended for another kernel/system.

jpoimboe commented:
Maybe we can look for CONFIG_LIVEPATCH in the .config file?

flaming-toast commented:
@jpoimboe That works (and is easier) too :-)

euspectre commented:
Good point. I will revisit the patch to check CONFIG_LIVEPATCH in the target's config instead.

euspectre commented:
Here is the updated patch.

jpoimboe commented on kpatch-build/kpatch-build:
```diff
@@ -97,24 +97,27 @@ find_dirs() {
 			SYMVERSFILE="$(readlink -f /lib/modules/$ARCHVERSION/extra/kpatch/Module.symvers)"
 		else
 			warn "unable to find Module.symvers for kpatch core module"
-			return 1
```
In this case I think we still want to return 1 so that it fails.

flaming-toast replied:
If we have `find_core_symvers` return 1 for an error, then I think we should move the `[[ -e "$SYMVERSFILE" ]]` check from line 454 to the very end of `find_core_symvers` and have it return 1 if the check fails. Then we could just do `find_core_symvers || die "can't find core module Module.symvers"`

euspectre replied:
Reasonable.

jpoimboe commented on kpatch-build/kpatch-build:
```diff
@@ -97,24 +97,27 @@ find_dirs() {
-		return
 	fi
-
-	return 1
```
Same here.

jpoimboe commented on kpatch-build/kpatch-build:
```diff
@@ -338,10 +341,7 @@ elif [[ $DISTRO = ubuntu ]] || [[ $DISTRO = debian ]]; then
-
-[[ -e "$SYMVERSFILE" ]] || die "can't find core module Module.symvers"
-
+find_dirs
```
|| die "can't find supporting tools"

flaming-toast commented on kpatch-build/kpatch-build:
```diff
@@ -97,24 +97,27 @@ find_dirs() {
 		DATADIR="$(readlink -f $SCRIPTDIR/../share/kpatch)"
+	fi
+}
+
```
Let's have `find_dirs()` return an error if it could not find `create-diff-object` in either location.
Then later when `find_dirs` is invoked, we can check for the error: 
`find_dirs || die "can't find supporting tools"`

euspectre replied:
Yes, that would be useful.

euspectre commented:
Here is v.3 of the patch.

As you suggested, find_dir() now returns error if create-diff-object is missing. find_core_symvers() now checks if the symvers file is actually present.

As for the returns in find_core_symvers(), I do not think they are necessary. In case of errors, SYMVERSFILE will remain empty and [[ -e "" ]] will return false. Added initialization of SYMVERSFILE to make things clearer.

I also removed one of the warnings about missing symvers file. One should be enough, I suppose.

jpoimboe commented:
Thanks @euspectre !

:+1:

flaming-toast commented:
Thanks @euspectre.

euspectre commented:
Thanks for the reviews! Three heads are much better than one ;-)
